iox_query/src/exec/gapfill.rs

```diff
@@ -604,4 +614,244 @@ mod test {
+        //     "| a  | 1970-01-01T00:00:01.125Z |    |",
+        //     "+----+--------------------------+----+",
+        // ];
+        // assert_batches_eq!(expected, &batches);
+        Ok(())
+    }
+
+    type ExprVec = Vec<Arc<dyn PhysicalExpr>>;
+
+    struct TestBatch {
+        group_cols: Vec<Vec<Option<&'static str>>>,
+        // Stored as millisecods since intervals use millis,
+        // to let test cases be consistent and easier to read.
+        time_col: Vec<Option<i64>>,
+        agg_cols: Vec<Vec<Option<i32>>>,
+    }
+
+    impl TestBatch {
+        fn num_cols(&self) -> usize {
+            self.group_cols.len() + self.agg_cols.len() + 1
+        }
+
+        fn schema(&self) -> SchemaRef {
+            let mut fields = Vec::with_capacity(self.num_cols());
+            for i in 0..self.group_cols.len() {
+                fields.push(Field::new(
+                    &format!("g{i}"),
+                    DataType::Dictionary(Box::new(DataType::Int32), Box::new(DataType::Utf8)),
+                    true,
+                ));
+            }
+            fields.push(Field::new(
+                "t",
+                DataType::Timestamp(TimeUnit::Nanosecond, None),
+                true,
+            ));
+            for i in 0..self.agg_cols.len() {
+                fields.push(Field::new(&format!("a{i}"), DataType::Int32, true));
+            }
+            Arc::new(Schema::new(fields))
+        }
+
+        fn exprs(&self) -> Result<(ExprVec, ExprVec)> {
+            let mut group_expr: ExprVec = vec![];
+            let mut aggr_expr: ExprVec = vec![];
+            let ngroup_cols = self.group_cols.len();
+            for i in 0..self.schema().fields().len() {
+                match i.cmp(&ngroup_cols) {
+                    Ordering::Less => group_expr.push(Arc::new(Column::new(&format!("g{i}"), i))),
+                    Ordering::Equal => group_expr.push(Arc::new(Column::new("t", i))),
+                    Ordering::Greater => {
+                        let idx = i - ngroup_cols + 1;
+                        aggr_expr.push(Arc::new(Column::new(&format!("a{idx}"), i)));
+                    }
+                }
+            }
+            Ok((group_expr, aggr_expr))
+        }
+    }
+
+    impl TryFrom<TestBatch> for RecordBatch {
+        type Error = DataFusionError;
+
+        fn try_from(value: TestBatch) -> Result<Self> {
+            let mut arrs: Vec<ArrayRef> =
+                Vec::with_capacity(value.group_cols.len() + value.agg_cols.len() + 1);
+            for gc in &value.group_cols {
+                let arr = Arc::new(DictionaryArray::<Int32Type>::from_iter(gc.iter().cloned()));
+                arrs.push(arr);
+            }
+            // Scale from milliseconds to the nanoseconds that are actually stored.
+            let scaled_times: Vec<Option<i64>> = value
+                .time_col
+                .iter()
+                .map(|o| o.map(|v| v * 1_000_000))
+                .collect();
```

Comment:
I think you can collect directly to the array too. Something like

```suggestion
            let arr: TimestampNanonsecondArray = value
                .time_col
                .iter()
                .map(|o| o.map(|v| v * 1_000_000))
                .collect();
```